src/sentry/static/sentry/app/components/contextPickerModal.jsx

```diff
@@ -0,0 +1,255 @@
+    };
+  }
+
+  componentDidMount() {
+    let {latestContext, organizations} = this.props;
+
+    // Don't make any assumptions if there are multiple organizations
+    if (organizations.length !== 1) {
+      return;
+    }
+
+    // If there is an org in context (and there's only 1 org available),
+    // attempt to see if we need more info from user and redirect otherwise
+    if (latestContext.organization) {
+      // This will handle if we can intelligently move the user forward
+      this.checkShouldFinish(
+        [latestContext.organization],
+        latestContext.organization.projects
+      );
+      return;
+    }
+
+    // Since user belongs to only 1 org, we can set it as active and componentWillReceiveProps handle the rest
+    fetchOrganizationDetails(organizations[0].slug, {
+      setActive: true,
+      loadProjects: true,
+    });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // Should only check the case where there is no latestContext.organization and we're waiting
+    // for it to be set (after fetch in DidMount)
+    let {latestContext} = this.props;
+    if (
+      (!latestContext.organization &&
+        latestContext.organization !== nextProps.latestContext.organization) ||
+      (latestContext.organization &&
+        nextProps.latestContext.organization &&
+        latestContext.organization.slug !== nextProps.latestContext.organization.slug)
+    ) {
+      // Check if we can push the user forward w/o needing them to select anything
+      this.checkShouldFinish(
+        this.props.organizations,
+        nextProps.latestContext.organization.projects,
+        nextProps.latestContext.organization
+      );
+    }
+  }
+
+  // Performs checks to see if we need to prompt user
+  // i.e. When there is only 1 org and no project is needed or
+  // there is only 1 org and only 1 project (which should be rare)
+  checkShouldFinish = (
+    organizations,
+    projects,
+    latestOrg = this.props.latestContext && this.props.latestContext.organization
+  ) => {
+    let {needProject, onFinish, nextPath} = this.props;
+
+    // If there is only one org and we dont need a project slug, then call finish callback
+    if (!needProject) {
+      if (organizations.length !== 1) {
+        this.setState({loading: false});
+        return;
+      }
+
+      onFinish(
+        replaceRouterParams(nextPath, {
+          orgId: organizations[0].slug,
+        })
+      );
+    } else {
+      if (projects.length !== 1) {
+        this.setState({loading: false});
+        return;
+      }
+
+      let org = latestOrg;
+      if (!org && organizations.length === 1) {
+        org = organizations[0];
+      }
+
+      onFinish(
+        replaceRouterParams(nextPath, {
+          orgId: org.slug,
+          projectId: projects[0].slug,
+        })
+      );
+    }
+  };
+
+  handleSelectOrganization = value => {
+    if (!this.props.needProject) {
+      this.checkShouldFinish([{slug: value}], []);
```

Comment:
Won't `props` propagation as a result of fetching the organization trigger calling `checkShouldFinish` via `componentWillReceiveProps`? Do we need to repeat it here? (Or is it a user experience thing since you can call it faster? If it's the latter I'd write a comment.)

Reply:
It saves an API request (rather defers it to the component we're navigating to).